# L2 Escape Models

The following models take into account that, throughout analyses, *no_goal* seems to behave entirely differently to the other two Goal Type conditions (*goal_frequent* and *goal_non_frequent*). 

Previous focus structure effects (see notebooks 11 to 6) as well as furhter exhaustive modelling (see notebooks 11_appendix vs. 12_appendix) show consistently the heterogeneity of combined data; by spliting data into goal contexts and no-goal contexts, both effects as well as models improved despite the pruning of observations. No-goal contexts are also analysed separately ((notebook 11)); however, and although results are coherent with what is seen by comparing combined data to goal behaviour data, the sample is clearly not enough to show anything significant from no-goal contexts.

Taking into account the effects summary (notebook 5) and response opportunity analysis (notebook 6), starting from the main effect identified in the previous analyses, predictors were added only when supported by opportunity.

Import Libraries

In [172]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Goal Behaviour Subgroup

1. Create escape_L2 
2. Counts and Proportions 
3. Logistic Models

     Main Effect Model:

    - E02: Agent

     Additive Models:

    - E04: Focus + Agent
    - E06: Agent + Goal_Type
    - E07: Focus + Agent + Goal_Type

4. Summary

Read Data

In [173]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create escape_2

Binary Outcome: responses that are not "L2_other": those who managed to escape L2_other by choosing either a *correct* response or an *L1_transfer* response.

In [174]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

Split data into goals vs. no_goals

In [175]:
goals = subgroup_theoretical[subgroup_theoretical["Goal_Type"]!= "no_goal"].copy()

In [176]:
no_goal = subgroup_theoretical[subgroup_theoretical["Goal_Type"]== "no_goal"].copy()

2. Counts and Proportions for goals and no_goal.

Escape_L2 responses count for goals:

In [177]:
goal_escape_L2_counts = pd.crosstab(
    goals["escape_L2"],
    goals["Response_Full"],
    margins= True)

goal_escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response,All
escape_L2,,,,,
0,0,26,0,0,26
1,62,0,43,1,106
All,62,26,43,1,132


Escape_L2 responses proportions for goals:

In [178]:
goals_escape_L2_props = goals["escape_L2"].value_counts(normalize = True)
goals_escape_L2_props

escape_L2
1    0.80303
0    0.19697
Name: proportion, dtype: float64

From 132 observations in the goal group, 80.3% of responses managed to escape L2_other options. 

Counts and Proportions by Condition:

In [179]:
goals_counts_condition = pd.crosstab(
    [goals["Goal_Type"], goals["Agent"], goals["Focus"]],
    goals["escape_L2"],
    margins = True)

goals_counts_condition

escape_L2                       0    1  All
Goal_Type         Agent Focus              
goal_frequent     0     I       8   16   24
                        They    6   12   18
                  1     I       2   14   16
                        They    0    8    8
goal_non_frequent 0     I       8   16   24
                        They    1   17   18
                  1     I       1   15   16
                        They    0    8    8
All                            26  106  132

In [180]:
goal_props_condition = pd.crosstab(
    [goals["Goal_Type"], goals["Agent"], goals["Focus"]],
    goals["escape_L2"],
    normalize= "index")

goal_props_condition

escape_L2                             0         1
Goal_Type         Agent Focus                    
goal_frequent     0     I      0.333333  0.666667
                        They   0.333333  0.666667
                  1     I      0.125000  0.875000
                        They   0.000000  1.000000
goal_non_frequent 0     I      0.333333  0.666667
                        They   0.055556  0.944444
                  1     I      0.062500  0.937500
                        They   0.000000  1.000000

Within goals, Agency seems to assist in the process of escaping the L2_other options.

- As previously mentioned, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space.

3. Logistic Models

Import Libraries

In [181]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
from scipy.stats import chi2


Sanity checks:

In [182]:
goals["escape_L2"].value_counts()

escape_L2
1    106
0     26
Name: count, dtype: int64

In [183]:
no_goal["escape_L2"].value_counts()

escape_L2
1    54
0    12
Name: count, dtype: int64

**Goal Contexts**

**Main Effect Models**

**E02**

escape_L2 ~ Agent

In [184]:
E02 = smf.logit(
    "escape_L2 ~ Agent",
    data=goals
    ).fit()

print(E02.summary())

Optimization terminated successfully.
         Current function value: 0.458568
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      130
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.07579
Time:                        20:18:21   Log-Likelihood:                -60.531
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001628
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.9754      0.245      3.986      0.000       0.496       1.455
Agent          1.7327      0.

Agency accounts for most of the explainable variation. 

**Additive Models**

**E04**

escape_L2 ~ Focus + Agent

In [185]:
E04 = smf.logit(
    "escape_L2 ~  Focus + Agent",
    data=goals
    ).fit()

print(E04.summary())

Optimization terminated successfully.
         Current function value: 0.446122
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      129
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                  0.1009
Time:                        20:18:21   Log-Likelihood:                -58.888
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001351
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.6432      0.298      2.156      0.031       0.058       1.228
Focus[T.They]     0.

Once Focus is added to the model with Agency, while it is still non-significant (p-value 0 0.080), it is much more informative than before. Focus explains additional variation in the data.

The whole model's LLR p-value (0.00135) is also better than agent alone (0.00163); not a lot but not requiring an interaction.

This model explains a lot of variation and demostrates parsimony.In addition, it shows coherent with directional effects across conditions from the opportunity analysis.

**E06**

escape_L2 ~ Agent + Goal_Type

In [186]:
E06= smf.logit(
    "escape_L2 ~  Agent + Goal_Type",
    data=goals
    ).fit()

print(E06.summary())

Optimization terminated successfully.
         Current function value: 0.451528
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      129
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.08998
Time:                        20:18:21   Log-Likelihood:                -59.602
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.002758
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.6854      0.317      2.162     

This is a good model even though the goal_non_frequent's p is still not siginificant, it captures another descriptive pattern. 

Non-frequent goals tend to suppress compteing L2 strategies more strongly. The coefficient points in precisely the direction predicted by the descriptive analysis.

**E07**

escape_L2 ~ Focus + Agent + Goal_Type

In [187]:
E07= smf.logit(
    "escape_L2 ~  Focus + Agent + Goal_Type",
    data=goals
    ).fit()

print(E07.summary())

Optimization terminated successfully.
         Current function value: 0.438864
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      128
Method:                           MLE   Df Model:                            3
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                  0.1155
Time:                        20:18:21   Log-Likelihood:                -57.930
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001709
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.3380      0.368      0.918     

Goal Type's p-value improves a little bit and the coefficient increases. The whole model is better (LLR p-value: 0.001709 vs. 0.002758) although not better than the focus-agency additive model's p (0.001351). Everything points in the same direction as in the opportunity analysis notebook.

Yet, this model shows a better picture of how counterfactual processing is shaped by considering the 3 cognitive dimensions (Focus, Agency and Goal_Type). In this sense, Agency shows a strong contribution, Focus shows a consistent positive contribution approaching significance and Goal Type shows a modest positive contribution. Focus and Goal Type still appear to move behaviour in systematic directions. 

4. Summary

- Separating the no-goal condition reduced heterogeneity in the remaining data, resulting in clearer predictor hierarchies coherent with opportunity analysis.

- Agency predicts escaping alternative L2 responses among participants who actually have a goal to protect. Agency opens the opportunity. Focus modifies where responses tend to go.  

- Goal Type consistently exhibited positive coefficients across additive models, although its independent contribution remained modest and did not reach conventional levels of statistical significance.

- Given results of the full additive model, escape responses are best understood as the joint product of agency, perspective, and goal structure, although agency provides the strongest independent contribution. Adding Goal Type preserves the pattern of effects and yields a coherent cognitive account in which agency remains the dominant predictor while focus and goal structure contribute more modestly.